### PR TITLE
OLM-2695: Adds downstream only CSV Namespace Labeler Plug-In

### PR DESCRIPTION
**Pull request by perdasilva**
#### Description
This PR adds the downstream only csv namespace labeler olm controller plug-in. This plug-in adds an additional csv queue informer with a sync method that when a csv is updated/created labels the csv namespace with `security.openshift.io/scc.podSecurityLabelSync=true` if:
 - the namespace is not part of the payload, unless it is `openshift-operators`
 - the namespace is not already labelled with the label above (whether true or false)
 - the namespace name is prefixed with `openshift-`

The csv informer ignores copied csvs and prunes out the body of the csv (save for the metadata section) to be frugal on memory.

**NOTE**: because this is a downstream only change, it makes changes to the staging directory. Because of this the `verify` job will fail - as it checks that there are no staging directory changes outside of what is brought in from the upstream.

### CODE REVIEW

File: staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	"k8s.io/client-go/util/workqueue"
+)
+
+const NamespaceLabelSyncerLabelKey = "security.openshift.io/scc.podSecurityLabelSync"
+
+// csvNamespaceLabelerPlugin is responsible for labeling non-payload openshift-* namespaces
+// with the label "security.openshift.io/scc.podSecurityLabelSync=true" so that the  PSA Label Syncer
+// see https://github.com/openshift/cluster-policy-controller/blob/master/pkg/psalabelsyncer/podsecurity_label_sync_controller.go
+// can help ensure that the operator payloads in the namespace continue to work even if they don't yet respect the
+// upstream Pod Security Admission controller, which will become active in k8s 1.15.
```
Comment (jianzhangbjz) [inline staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go:30]: I guess it should be `1.25` here.

Comment (perdasilva) [reply staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go]: Great catch! I've fixed it =D

**Comment by jianzhangbjz** [comment]
Found an issue: An `openshift-` prefixed project's `security.openshift.io/scc.podSecurityLabelSync=false` cannot be updated to `security.openshift.io/scc.podSecurityLabelSync=true` after installing an operator. Details:
```yaml
MacBook-Pro:~ jianzhang$ oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.12.0-0.ci.test-2022-08-31-090742-ci-ln-m7pqtgb-latest   True        False         42m     Cluster version is 4.12.0-0.ci.test-2022-08-31-090742-ci-ln-m7pqtgb-latest

MacBook-Pro:~ jianzhang$ oc adm new-project openshift-test2 
Created project openshift-test2
MacBook-Pro:~ jianzhang$ 
MacBook-Pro:~ jianzhang$ oc label ns openshift-test2  security.openshift.io/scc.podSecurityLabelSync=false
namespace/openshift-test2 labeled

MacBook-Pro:~ jianzhang$ oc get sub -n openshift-test2
NAME   PACKAGE   SOURCE                CHANNEL
etcd   etcd      community-operators   singlenamespace-alpha
MacBook-Pro:~ jianzhang$ oc get ip -n openshift-test2
NAME            CSV                   APPROVAL    APPROVED
install-tk9r4   etcdoperator.v0.9.4   Automatic   true
MacBook-Pro:~ jianzhang$ oc get csv -n openshift-test2
NAME                  DISPLAY   VERSION   REPLACES              PHASE
etcdoperator.v0.9.4   etcd      0.9.4     etcdoperator.v0.9.2   Succeeded
MacBook-Pro:~ jianzhang$ oc get pods -n openshift-test2
NAME                             READY   STATUS    RESTARTS   AGE
etcd-operator-76bc8ff57d-7mpcx   3/3     Running   0          2m54s

MacBook-Pro:~ jianzhang$ oc get ns openshift-test2 -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/sa.scc.mcs: s0:c26,c25
    openshift.io/sa.scc.supplemental-groups: 1000700000/10000
    openshift.io/sa.scc.uid-range: 1000700000/10000
  creationTimestamp: "2022-08-31T10:03:54Z"
  labels:
    kubernetes.io/metadata.name: openshift-test2
    olm.operatorgroup.uid/b9b3218c-95a4-48f3-be89-b249838396af: ""
    security.openshift.io/scc.podSecurityLabelSync: "false"
  name: openshift-test2
  resourceVersion: "50036"
  uid: 4b073873-b359-4cdc-a6d1-703f003e0ea2
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

**Comment by perdasilva** [comment]
> Found an issue: An `openshift-` prefixed project's `security.openshift.io/scc.podSecurityLabelSync=false` cannot be updated to `security.openshift.io/scc.podSecurityLabelSync=true` after installing an operator. Details:

This should be the correct behaviour. This just means the namespace has opted-out of the label sync'er behavior entirely. The plug-in will ignore namespaces that already contain the label (with either value).

@joelanford correct me if I'm wrong, if the plug-in labels the namespace, and the user deletes the label, are we expected to re-label it? or do we not offer any guarantees after that?

**Comment by jianzhangbjz** [comment]
> This should be the correct behaviour. This just means the namespace has opted-out of the label sync'er behavior entirely. The plug-in will ignore namespaces that already contain the label (with either value).

Make sense, thanks!

> if the plug-in labels the namespace, and the user deletes the label, are we expected to re-label it? 

I guess it should be re-labeled, and I retest it. It works as expected. Details,
```yaml
MacBook-Pro:~ jianzhang$ oc label ns openshift-test2  security.openshift.io/scc.podSecurityLabelSync-
namespace/openshift-test2 unlabeled
MacBook-Pro:~ jianzhang$ oc get ns  openshift-test2  -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/sa.scc.mcs: s0:c26,c10
    openshift.io/sa.scc.supplemental-groups: 1000670000/10000
    openshift.io/sa.scc.uid-range: 1000670000/10000
  creationTimestamp: "2022-09-01T09:55:35Z"
  labels:
    kubernetes.io/metadata.name: openshift-test2
    olm.operatorgroup.uid/6e6e3a40-4bca-48ff-867e-1034686892ca: ""
  name: openshift-test2
  resourceVersion: "32213"
  uid: 693f511c-8594-4dc0-b14c-43f6e4454e02
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```
And, after a while, it was re-labeled.
```yaml
MacBook-Pro:~ jianzhang$ oc get ns  openshift-test2  -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/sa.scc.mcs: s0:c26,c10
    openshift.io/sa.scc.supplemental-groups: 1000670000/10000
    openshift.io/sa.scc.uid-range: 1000670000/10000
  creationTimestamp: "2022-09-01T09:55:35Z"
  labels:
    kubernetes.io/metadata.name: openshift-test2
    olm.operatorgroup.uid/6e6e3a40-4bca-48ff-867e-1034686892ca: ""
    pod-security.kubernetes.io/enforce: restricted
    pod-security.kubernetes.io/enforce-version: v1.24
    security.openshift.io/scc.podSecurityLabelSync: "true"
  name: openshift-test2
  resourceVersion: "32336"
  uid: 693f511c-8594-4dc0-b14c-43f6e4454e02
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```
 One more question, after uninstalling the operator, will the label be removed? Thanks! (PS: based on my understanding, no need, but need your confirm)

- Uninstall the operator

```yaml
MacBook-Pro:~ jianzhang$ oc get csv -n  openshift-test2
No resources found in openshift-test2 namespace.
```

-  The label is still there after a while.

```yaml
MacBook-Pro:~ jianzhang$ oc get ns  openshift-test2  -o yaml
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    openshift.io/description: ""
    openshift.io/display-name: ""
    openshift.io/sa.scc.mcs: s0:c26,c10
    openshift.io/sa.scc.supplemental-groups: 1000670000/10000
    openshift.io/sa.scc.uid-range: 1000670000/10000
  creationTimestamp: "2022-09-01T09:55:35Z"
  labels:
    kubernetes.io/metadata.name: openshift-test2
    olm.operatorgroup.uid/6e6e3a40-4bca-48ff-867e-1034686892ca: ""
    pod-security.kubernetes.io/enforce: restricted
    pod-security.kubernetes.io/enforce-version: v1.24
    security.openshift.io/scc.podSecurityLabelSync: "true"
  name: openshift-test2
  resourceVersion: "32336"
  uid: 693f511c-8594-4dc0-b14c-43f6e4454e02
spec:
  finalizers:
  - kubernetes
status:
  phase: Active
```

**Comment by perdasilva** [comment]
@jianzhangbjz I've just fixed up the implementation to ensure that the namespace will be relabelled. In the old implementation the namespace *would* be relabelled but only when a csv add/update event happened. The new implementation watches namespaces for namespace events. I've also revamped the implementation to be more selective about the event sources, cutting down on the number of time the sync method would be called.

For your original question, it's no necessary to unlabel the namespace if all csvs are gone :+1:

**Comment by jianzhangbjz** [comment]
Hi @perdasilva , thanks for your confirmation! I retest this feature PR by following [OCP-53914](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-53914), and all steps passed. And, could you help review that test case and add the `tc-approved` label for https://issues.redhat.com/browse/OLM-2695? Thanks!

**Comment by jianzhangbjz** [comment]
LGTM, 
/unhold

**Review by timflannagan — APPROVED** [review]

File: staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+		// ignore namespaces that are *NOT* prefixed with openshift- but accept metav1.NamespaceAll
+		if !(hasOpenshiftPrefix(namespace)) && namespace != metav1.NamespaceAll {
+			continue
+		}
+
+		nonCopiedCsvInformer := newNonCopiedCsvInformerForNamespace(namespace, config.ExternalClient(), config.ResyncPeriod()())
+
+		nonCopiedCsvQueue := workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("%s/csv-ns-labeler-plugin-csv-queue", namespace),
```
Comment (bentito) [inline staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go:101]: shouldn't this be plugin.log() or something?

Comment (perdasilva) [reply staging/operator-lifecycle-manager/pkg/controller/operators/olm/plugins/downstream_csv_namespace_labeler_plugin.go:101]: this isn't a log statement, it's the name of the queue - just using string formatting to build it

**Comment by perdasilva** [comment]
verify is failing due to changes in the staging directory. This is expected since the downstream only changes are made in there.

**Review by timflannagan — APPROVED** [review]

**Pull request merged**